src/accountsdb/buffer_pool.zig

```diff
@@ -735,130 +803,322 @@ pub const HierarchicalFIFO = struct {
-            return self.cached_read.readByte(self.bytes_read);
-        }
+    /// Only called by the BufferPool
+    fn initCached(
+        buffer_pool: *BufferPool,
+        frame_indices: []const FrameIndex,
+        first_frame_start_offset: FrameOffset,
+        last_frame_end_offset: FrameOffset,
+    ) ReadHandle {
+        return .{
+            .cached = .{
+                .buffer_pool = buffer_pool,
+                .frame_indices = frame_indices,
+                .first_frame_start_offset = first_frame_start_offset,
+                .last_frame_end_offset = last_frame_end_offset,
+            },
+        };
+    }
 
-        pub fn reset(self: *Iterator) void {
-            self.bytes_read = 0;
-        }
+    /// External to the BufferPool, data will be freed upon .deinit
+    pub fn initAllocatedOwned(data: []u8) ReadHandle {
+        return ReadHandle{
+            .allocated_read = .{ .slice = data, .owned = true },
+        };
+    }
 
-        pub fn readBytes(self: *Iterator, buffer: []u8) error{}!usize {
-            var i: u32 = 0;
-            while (i < buffer.len) : (i += 1) {
-                buffer[i] = self.next() orelse break;
-            }
-            return i;
-        }
+    /// External to the BufferPool
+    pub fn initAllocated(data: []u8) ReadHandle {
+        return ReadHandle{
+            .allocated_read = .{ .slice = data, .owned = false },
+        };
+    }
 
-        pub fn reader(self: *Iterator) Reader {
-            return .{ .context = self };
+    pub fn deinit(self: ReadHandle, allocator: std.mem.Allocator) void {
+        switch (self) {
+            .cached => |cached| {
+                for (cached.frame_indices) |frame_index| {
+                    std.debug.assert(frame_index != INVALID_FRAME);
+
+                    if (cached.buffer_pool.frames_metadata.rc[frame_index].release()) {
+                        // notably, the frame remains in memory, and its hashmap entry
+                        // remains valid.
+                    }
+                }
+                allocator.free(cached.frame_indices);
+            },
+            .sub_read => |_| {},
+            .allocated_read => |allocated_read| {
+                if (allocated_read.owned) allocator.free(allocated_read.slice);
+            },
         }
-    };
+    }
+
+    pub fn iterator(self: *const ReadHandle) Iterator {
+        return .{ .read_handle = self, .start = 0, .end = self.len() };
+    }
 
-    pub fn readByte(self: CachedRead, index: usize) u8 {
-        std.debug.assert(self.frame_indices.len != 0);
-        std.debug.assert(index < self.len());
-        const offset = index + self.first_frame_start_offset;
-        std.debug.assert(offset >= self.first_frame_start_offset);
+    /// Copies entire read into specified buffer. Must be correct length.
+    pub fn readAll(self: ReadHandle, buf: []u8) error{InvalidArgument}!void {
+        try self.read(0, self.len(), buf);
+    }
```

Comment:
In tandem with my other suggestion concerning `read`, would then need to adjust this to explicitly check `self.len() != buf.len`.

Reply:
26cd92e81d02dd6073b05ab15975611a7d696393